dpcpp/solver/idr_kernels.dp.cpp

```diff
@@ -55,11 +66,760 @@ namespace dpcpp {
+                         sycl::nd_item<3> item_ct1)
+{
+    const auto global_id = thread::get_thread_id_flat(item_ct1);
+    const auto row = global_id / m_stride;
+    const auto col = global_id % m_stride;
+
+    if (global_id < nrhs) {
+        stop_status[global_id].reset();
+    }
+
+    if (row < subspace_dim && col < nrhs * subspace_dim) {
+        m_values[row * m_stride + col] =
+            (row == col / nrhs) ? one<ValueType>() : zero<ValueType>();
+    }
+}
+
+template <typename ValueType>
+void initialize_m_kernel(dim3 grid, dim3 block, size_t dynamic_shared_memory,
+                         sycl::queue *stream, size_type subspace_dim,
+                         size_type nrhs, ValueType *m_values,
+                         size_type m_stride, stopping_status *stop_status)
+{
+    stream->submit([&](sycl::handler &cgh) {
+        cgh.parallel_for(
+            sycl_nd_range(grid, block), [=](sycl::nd_item<3> item_ct1) {
+                initialize_m_kernel(subspace_dim, nrhs, m_values, m_stride,
+                                    stop_status, item_ct1);
+            });
+    });
+}
+
+
+template <size_type block_size, typename ValueType>
+void orthonormalize_subspace_vectors_kernel(
+    size_type num_rows, size_type num_cols, ValueType *__restrict__ values,
+    size_type stride, sycl::nd_item<3> item_ct1,
+    UninitializedArray<ValueType, block_size> *reduction_helper_array,
+    remove_complex<ValueType> *reduction_helper_real)
+{
+    const auto tidx = thread::get_thread_id_flat(item_ct1);
+
+
+    ValueType *__restrict__ reduction_helper = (*reduction_helper_array);
+
+
+    for (size_type row = 0; row < num_rows; row++) {
+        for (size_type i = 0; i < row; i++) {
+            auto dot = zero<ValueType>();
+            for (size_type j = tidx; j < num_cols; j += block_size) {
+                /*
+                DPCT1007:5: Migration of this CUDA API is not supported by the
+                Intel(R) DPC++ Compatibility Tool.
+                */
+                dot += values[row * stride + j] * conj(values[i * stride + j]);
+            }
+            // TODO: check with intel why we need this here.
+            item_ct1.barrier();
+            reduction_helper[tidx] = dot;
+
+            /*
+            DPCT1065:3: Consider replacing sycl::nd_item::barrier() with
+            sycl::nd_item::barrier(sycl::access::fence_space::local_space) for
+            better performance, if there is no access to global memory.
+            */
+            item_ct1.barrier();
+            ::gko::kernels::dpcpp::reduce(
+                group::this_thread_block(item_ct1), reduction_helper,
+                [](const ValueType &a, const ValueType &b) { return a + b; });
+            /*
+            DPCT1065:4: Consider replacing sycl::nd_item::barrier() with
+            sycl::nd_item::barrier(sycl::access::fence_space::local_space) for
+            better performance, if there is no access to global memory.
+            */
```

Comment:
DPCT issue. Since we use the `local` accessor, I think the suggestion might make sense?